modules/docs/arrow-docs/docs/docs/arrow/typeclasses/traverse/README.md

```diff
@@ -11,16 +11,366 @@ redirect_from:
+
+Going back to our `Promise` example, we can get an `Applicative` instance for `IO`, by converting `Promise` to `IO` that runs each `Promise` concurrently. Then when we traverse a `List<A>` with an `(A) -> Promise<ForIO, B>`, we can imagine the traversal as a scatter-gather. Each `A` creates a concurrent computation that will produce a `B` (the scatter), and as the `Promise`s complete they will be gathered back into a `List`.
+
+{TODO: finish example with Promise}
+
+```kotlin:ank:playground
+import arrow.core.Either
+import arrow.core.Option
+import arrow.fx.Promise
+
+interface Profile
+interface User
+
+fun parseInt(s: String): Option<Int> = TODO()
+data class SecurityError(val message: String)
+data class Credentials(val email: String, val password: String)
+
+fun validateLogin(c: Credentials): Either<SecurityError, Unit> = TODO()
+
+fun userInfo(u: User): Promise<ForIO, Profile> = TODO()
+
+data class Admin(val name: String, val crd: Credentials): User
+
+fun ListK<User>.processLogin()
+```
+
+Evidently, `Traverse` is not limited to `List` or `Nel`, it provides an abstraction over 'things that can be traversed over', like a Binary tree, hence the name `Traverse`.
+
+#### Playing with `Reader`
+
+Another interesting effect we can use is Reader. Recall that a `Reader<D, A>` is a type alias for `Kleisli<ForId, D, A>` which is a wrapper around `(D) -> A`.
+
+If we fix `D` to be some sort of dependency or configuration, we can use the `Reader` applicative in our `traverse`.
+
+```kotlin:ank
+import arrow.mtl.Reader
+
+interface Context
+interface Topic
+interface Result
+
+typealias Job<A> = Reader<Context, A>
+
+fun processTopic(t: Topic): Job<Result> = TODO()
+```
+
+We can imagine we have a data pipeline that processes a bunch of data, each piece of data being categorized by a topic. Given a specific topic, we produce a `Job` that processes that topic. (Note that since a `Job` is just a `Reader`/`Kleisli`, one could write many small `Jobs` and compose them together into one `Job` that is used/returned by `processTopic`.)
+
+Corresponding to our bunches of data are bunches of topics, a `List<Topic>` if you will. Since Reader has an `Applicative` instance, we can `traverse` over this list with `processTopic`.
+
+```kotlin:ank
+import arrow.core.Id
+import arrow.core.ForId
+import arrow.core.ListK
+import arrow.core.extensions.id.applicative.applicative
+import arrow.mtl.KleisliPartialOf
+import arrow.mtl.extensions.kleisli.applicative.applicative
+import arrow.mtl.fix
+
+fun processTopics(topics: ListK<Topic>): Job<ListK<Result>> =
+  topics.traverse<KleisliPartialOf<ForId, Context>, Result>(Job.applicative<ForId, Context>(Id.applicative())) {
```

Comment:
I'd extract the instance `Job.applicative<ForId, Context>(Id.applicative())` to a field here.